usd/src/sdf_parser/Joint.cc

```diff
@@ -0,0 +1,402 @@
+
+#include <ignition/math/Angle.hh>
+#include <ignition/math/Pose3.hh>
+#include <ignition/math/Vector3.hh>
+
+// TODO(ahcorde) this is to remove deprecated "warnings" in usd, these warnings
+// are reported using #pragma message so normal diagnostic flags cannot remove
+// them. This workaround requires this block to be used whenever usd is
+// included.
+#pragma push_macro ("__DEPRECATED")
+#undef __DEPRECATED
+#include <pxr/base/gf/quatf.h>
+#include <pxr/base/gf/vec3f.h>
+#include <pxr/base/tf/token.h>
+#include <pxr/usd/sdf/path.h>
+#include <pxr/usd/usd/relationship.h>
+#include <pxr/usd/usd/stage.h>
+#include <pxr/usd/usdPhysics/driveAPI.h>
+#include <pxr/usd/usdPhysics/fixedJoint.h>
+#include <pxr/usd/usdPhysics/joint.h>
+#include <pxr/usd/usdPhysics/prismaticJoint.h>
+#include <pxr/usd/usdPhysics/revoluteJoint.h>
+#include <pxr/usd/usdPhysics/sphericalJoint.h>
+#pragma pop_macro ("__DEPRECATED")
+
+#include "sdf/Error.hh"
+#include "sdf/Joint.hh"
+#include "sdf/JointAxis.hh"
+#include "sdf/Link.hh"
+#include "sdf/Model.hh"
+#include "../UsdUtils.hh"
+
+namespace sdf
+{
+// Inline bracke to help doxygen filtering.
+inline namespace SDF_VERSION_NAMESPACE {
+//
+namespace usd
+{
+  /// \brief Helper function for setting a USD joint's pose relative to the
+  /// joint's parent and child.
+  /// \param[in] _jointPrim The USD joint prim
+  /// \param[in] _joint The SDF representation of _jointPrim
+  /// \param[in] _parentModel The SDF model that is the parent of _joint
+  /// \return UsdErrors, which is a vector of UsdError objects. Each UsdError
+  /// includes an error code and message. An empty vector indicates no error
+  /// occurred when setting _joint's pose relative to _joint's parent and child
+  UsdErrors SetUSDJointPose(pxr::UsdPhysicsJoint &_jointPrim,
+      const sdf::Joint &_joint, const sdf::Model &_parentModel)
+  {
+    UsdErrors errors;
+
+    ignition::math::Pose3d parentToJoint;
+    if (_joint.ParentLinkName() == "world")
+    {
+      ignition::math::Pose3d modelToJoint;
+      auto poseErrors = usd::PoseWrtParent(_joint, modelToJoint);
+      if (!poseErrors.empty())
+        return poseErrors;
+
+      // it is assumed the _parentModel's parent is the world
+      ignition::math::Pose3d worldToModel;
+      poseErrors = usd::PoseWrtParent(_parentModel, worldToModel);
+      if (!poseErrors.empty())
+        return poseErrors;
+
+      parentToJoint = worldToModel * modelToJoint;
+    }
+    else
+    {
+      auto poseResolutionErrors =
+        _joint.SemanticPose().Resolve(parentToJoint, _joint.ParentLinkName());
+      if (!poseResolutionErrors.empty())
+      {
+        errors.push_back(UsdError(
+              sdf::Error(sdf::ErrorCode::POSE_RELATIVE_TO_INVALID,
+              "Unable to get the pose of joint [" + _joint.Name() +
+              "] w.r.t. its parent link [" + _joint.ParentLinkName() + "].")));
+        for (const auto &e : poseResolutionErrors)
+          errors.push_back(UsdError(e));
```

Comment:
This case is a little different than other error cases, because `poseResolutionErrors` holds a list of `sdf::Error` objects, but `errors` is of type `UsdError`. So, as you can see in the loop, we have to construct a `UsdError` object from the existing `sdf::Error` object. Since we have to create a different type of error object before adding it to the error list, I think we will need to keep this loop.

Reply:
got it